### PR TITLE
fix: connection and consumer recovery

### PR DESCRIPTION
**Pull request by wz185**

### CODE REVIEW

File: src/rabbitmq/RabbitmqClient.ts [hunk 1]
```diff
@@ -71,49 +71,63 @@ export abstract class RabbitmqClient {
...
+   * 1. Reconnect when errors occur.
+   * 2. Errors do not exist if connection.close() is called
+   *  or a server initiated graceful close.
+   * 3. Graceful closed connection will be recovered.
+   * 4. Because close event with an error will be emitted after connection error event,
+   *  only handle connection close event.
    * @param err
    */
-  protected handleConnectionClose(err: Error) {
+  protected async handleConnectionClose(err: Error) {
```
Comment (willowyang80) [inline src/rabbitmq/RabbitmqClient.ts]: should be `protected async handleConnectionClose(err?: Error) {`

File: src/rabbitmq/RabbitmqClient.ts [hunk 2]
```diff
@@ -123,10 +137,17 @@ export abstract class RabbitmqClient {
...
+   *  1. failed to consume.
+   *
+   * Channel error event will trigger connection error event which will
+   * trigger connection close event. Do not handle connection error event because a close event with error will be emiited.
+   * Then the close event will be handled.
    */
-  protected handleError(err: Error) {
-    this.logger.error(err, this.debugMsg(`Handling connection or channel error. Will reconnect.`));
-    this.closeAllAndScheduleReconnection();
+  protected async handleError(tag: string, err: Error) {
```
Comment (willowyang80) [inline src/rabbitmq/RabbitmqClient.ts]: tag should be an enum variable

File: src/rabbitmq/RabbitmqClient.ts [hunk 3]
```diff
@@ -140,57 +161,69 @@ export abstract class RabbitmqClient {
...
     this.logger.error(this.debugMsg(`Couldn't reconnect after ${this.clientConfig.maxConnectionAttempts} attempts`));
 
-    this.reconnecting = false;
+    this.stopReconnecting();
     // tslint:disable-next-line
     if (this.clientConfig.exitOnIrrecoverableReconnect !== false) {
       this.logger.error(this.debugMsg('Cowardly refusing to continue. Calling shutdown function'));
       this.shutdownFunction();
     }
+    return Promise.resolve(false);
```
Comment (willowyang80) [inline src/rabbitmq/RabbitmqClient.ts]: just return false. don't wrap the result with another promise.

File: src/rabbitmq/RabbitmqConsumer.ts [hunk 4]
```diff
@@ -92,14 +93,16 @@ export class RabbitmqConsumer extends RabbitmqClient {
...
     }
-    return this.channel.consume(
+    const rc: RepliesConsume = await this.channel.consume(
       queueName,
       (message: Message) => {
           this.handleMessage(message);
       },
       consumerConfig.options);
+    this.consumerTag = rc.consumerTag;
+    return Promise.resolve(rc);
```
Comment (willowyang80) [inline src/rabbitmq/RabbitmqConsumer.ts]: just return rc

File: src/util/PromiseUtil.ts [hunk 5]
```diff
@@ -24,10 +24,8 @@ export abstract class PromiseUtil {
     }
     return p.then(() => result);
   }
-  static sleepPromise<H>(ms, returnVal?: H): Promise<H> {
-    return new Promise<H>((resolve) => {
-      setTimeout(() => { resolve(returnVal); }, ms);
-    });
+  static async sleepPromise(ms): Promise<any> {
```
Comment (willowyang80) [inline src/util/PromiseUtil.ts]: better keep the generic type here 

`static sleepPromise<H>(ms): Promise<H> {`

Comment (wz185) [reply src/util/PromiseUtil.ts]: why? type is not used.

File: src/rabbitmq/RabbitmqClient.ts [hunk 6]
```diff
@@ -71,49 +76,63 @@ export abstract class RabbitmqClient {
...
-    this.reconnecting = false;
+  protected async connect(addHandler = true): Promise<void> {
+    await this.createConnection(addHandler);
+    await this.createChannel(addHandler);
+    this.stopReconnecting();
     this.readyGate.channelReady();
   }
 
-  private async createConnection() {
+  private async createConnection(addHandler: boolean): Promise<void> {
```
Comment (willowyang80) [inline src/rabbitmq/RabbitmqClient.ts]: need a default value for addHandler here just in case this being called somewhere else?

**Pull request merged**